Item refactor

elduende05 opened the pull request:
1st batch of commits:
- all missing materials and key items
- animated icons with menu option
- localization of item descriptions
- item registration cleanup

2nd batch of commits:
- all remaining missing weapons, bows, shields, armor, and food
- finish decoupling item enum so list is based off json
- refactor item icons into subdirectories

iTNTPiston commented on src/core/Item.ts:
```diff
@@ -28,55 +31,261 @@ export type ItemStack = {
+    NormalArrow = "NormalArrow",
+
+    // Deprecated Materials
     Lotus = "Lotus",
```
We don't really want to allow duplicated enums for the same material, since users may take advantage of this, and we will break existing scripts when we remove this.

elduende05 replied:
Should I just leave the enum registered with the short names like "Lotus" for now but convert its existence elsewhere (image file name)?

elduende05 replied:
I took the above approach.

iTNTPiston commented on src/core/Item.ts:
```diff
@@ -97,96 +308,38 @@ const TypeToCount = {
-register(9, Item.ZoraArmor, ItemType.Armor, {
-	stackable: false
+itemList.items.forEach(itemData => {
+	// Default id to 0 if undefined in config file
```
Id is completely unused now. Can remove id from config file completely

elduende05 replied:
I'll go cull it from the config file and remove the associated code.

iTNTPiston commented on src/core/Item.ts:
```diff
@@ -97,96 +308,38 @@ const TypeToCount = {
 	ItemToData[item] = data;
 };
+const getItemTypeValueByName = (type: string) => {
+	return Object.entries(ItemType).find(([key]) => key === type)?.[1];
```
Can we use something clearer here? `ItemType[type]` should give you the enum value by name

elduende05 replied:
If something seems weird to you, it's probably my lack of familiarity with JS syntax in general or TS specifically. I'll see if what you suggested works.

elduende05 replied:
That expression does not appear to work.

elduende05 replied:
`(<any>ItemType)[type]` seems to work

iTNTPiston commented on src/core/Slots.ts:
```diff
@@ -115,7 +115,7 @@ export class Slots {
 		// Key item check: if the key item or master sword already exists in the first tab, do not add
 		if(mCount != 0){
-			if(data.type === ItemType.Key || item === Item.MasterSword) {
+			if(data.type === ItemType.Key || item === Item.MasterSword || item === Item.MasterSwordAwakened) {
```
Does the game treat Master Sword and Master Sword Awakened as 2 separate weapons? AFAIK the upgrade part is just a flag. If we treat them as separate, this can become buggy during the duplication check (or we have to add extra logic)

elduende05 replied:
I'm not actually sure. Same with enabled / disabled versions of the champion abilities. Would require testing in game or perhaps asking leoetlino or kinak in the discord. Maybe for these we could leave the icons in for now, remove the item registration for them, then come back later to add options to mark the MS as awakened or not and enable / disable each champ ability?

iTNTPiston replied:
Sure

iTNTPiston commented on src/App.tsx:
```diff
@@ -33,6 +33,7 @@ export const App: React.FC =  () => {
 	const [page, setPageInState] = useState<string>("#simulation");
 	// Option States
 	const [interlaceInventory, setInterlaceInventory] = useState<boolean>(false);
+	const [animatedIcons, setAnimatedIcons] = useState<boolean>(false);
```
nit - can we call this `isIconAnimated` to indicate it's a flag?

elduende05 replied:
Sure thing.

iTNTPiston commented on src/config/i18n/en_US.json:
```diff
@@ -0,0 +1,246 @@
+{
```
Did you generate this from yaml? If so, can include the yaml source since we can use a webpack loader to load yaml files directly in the future

elduende05 replied:
I added this directly as json by grabbing the list of items, regexing it to roughly the form that was needed, then going in and adding spaces / formatting the display values as appropriate.

iTNTPiston reviewed (changes requested):
See my comments.

elduende05 commented on src/core/Localization.ts:
```diff
@@ -0,0 +1,6 @@
+import display from "config/i18n/en_US.json";
+
+export const getDisplayValue = (displayKey: string, defaultValue?: string) => {
+	return Object.entries(display).find(([key]) => key === displayKey)?.[1] || defaultValue || displayKey;
```
Lack of familiarity with syntax. I'll try what you suggested.

elduende05 commented on src/core/DisplayableInventory.ts:
```diff
@@ -1,7 +1,10 @@
 
 export type DisplayableSlot = {
     image: string,
+    animatedImage?: string,
```
Figured something out, although the amount of code change seems to be about the same since I have to pass the animatedIcons option around a bit more up top.

elduende05 commented:
I think I've addressed all your concerns. I kept the review feedback separate for now, but if you think this looks ok I can squash the changes down into the appropriate commits.

iTNTPiston commented on src/core/Item.ts:
```diff
@@ -323,26 +294,16 @@ const register = (_id: number, item: Item, type: ItemType, options?: Partial<Ite
-	// Default id to 0 if undefined in config file
-	const id = itemData.id ? Number(itemData.id) : 0;
-	register(id, itemData.item as Item, getItemTypeValueByName(itemData.type) as ItemType, itemData.options);
+	register(itemData.item as Item, (<never>ItemType)[itemData.type], itemData.options);
```
Generally it's a bad idea to use `never` or `any`. The linter is going to complain. One of the ways to solve casting is have a type declaration for the entire `itemList` structure, like `type ItemDataObject = { ... }`. This way we also get some documentation of what the json type is supposed to be

elduende05 replied:
I see. I can try to put something like that in place. Certainly nice to have some type safety.

iTNTPiston commented on src/core/Localization.ts:
```diff
@@ -2,5 +2,5 @@
 
 export const getDisplayValue = (displayKey: string, defaultValue?: string) => {
-	return Object.entries(display).find(([key]) => key === displayKey)?.[1] || defaultValue || displayKey;
+	return (<never>display)[displayKey] ?? defaultValue ?? displayKey;
```
`type LanguageMap = Record<string, string>;`, then you can cast `display` to that and not have wacky types

elduende05 replied:
Nice. I'll try that.

elduende05 commented on src/App.css:
```diff
@@ -86,6 +86,7 @@ h1 {
   border: 1px solid #999999;
   box-sizing: border-box;
   margin:1px;
+  padding:1px;
```
The items actually have a small amount of padding in game and some of the icons looked weird stretched all the way to the border (the saddles and bridles mostly). Try removing it and comparing those last few key items to see the effect, but it was actually true for all items throughout the inventory. Originally I started updating the images but then realized you kinda don't want to pre-bake your padding if you don't need to. It would have also taken forever to do that for all 500+ icons.
